c/tests/test_tables.c

```diff
@@ -4760,47 +4760,176 @@ test_link_ancestors_multiple_to_single_tree(void)
 }
 
 static void
-test_ibd_finder(void)
+verify_ibd_segment_list(tsk_segment_list_t *list, tsk_size_t num_nodes)
+{
+    tsk_segment_t *seg;
+    double total_span = 0;
+    tsk_size_t num_segments = 0;
+    /* double last_right = 0; */
+
+    for (seg = list->head; seg != NULL; seg = seg->next) {
+        CU_ASSERT_FATAL(seg->left < seg->right);
+        CU_ASSERT_FATAL(seg->node >= 0);
+        CU_ASSERT_FATAL(seg->node < (tsk_id_t) num_nodes);
+        total_span += seg->right - seg->left;
+        num_segments++;
+
+        /* TODO the segments are not necessarily in order - issue #1682 */
+        /* CU_ASSERT_FATAL(seg->left >= last_right); */
+        /* last_right = seg->right; */
+    }
+    CU_ASSERT_EQUAL_FATAL(total_span, list->total_span);
+    CU_ASSERT_EQUAL_FATAL(num_segments, list->num_segments);
+}
+
+static void
+verify_ibd_result(tsk_ibd_result_t *result)
+{
+    int ret;
+    tsk_size_t j;
+    tsk_id_t a, b;
+    int64_t index;
+    tsk_size_t total_segments = 0;
+    tsk_size_t num_pairs = tsk_ibd_result_get_num_pairs(result);
+    tsk_id_t *pairs
+        = tsk_malloc(2 * tsk_ibd_result_get_num_pairs(result) * sizeof(*pairs));
+    tsk_id_t *pairs2
+        = tsk_malloc(2 * tsk_ibd_result_get_num_pairs(result) * sizeof(*pairs));
+    tsk_segment_list_t **lists
+        = tsk_malloc(tsk_ibd_result_get_num_pairs(result) * sizeof(*lists));
+    tsk_avl_node_int_t **avl_nodes
+        = tsk_malloc(result->pair_map.size * sizeof(*avl_nodes));
+
+    CU_ASSERT_FATAL(pairs != NULL);
+    CU_ASSERT_FATAL(pairs2 != NULL);
+    CU_ASSERT_FATAL(avl_nodes != NULL);
+    CU_ASSERT_FATAL(lists != NULL);
+    CU_ASSERT_EQUAL_FATAL(num_pairs, result->pair_map.size);
+    tsk_ibd_result_print_state(result, _devnull);
+
+    ret = tsk_ibd_result_get_keys(result, pairs);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    ret = tsk_avl_tree_int_ordered_nodes(&result->pair_map, avl_nodes);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    for (j = 0; j < num_pairs; j++) {
+        a = pairs[2 * j];
+        b = pairs[2 * j + 1];
+        index = a * (int64_t) result->num_nodes + b;
+        CU_ASSERT(a < b);
+        CU_ASSERT_EQUAL(tsk_avl_tree_int_search(&result->pair_map, index), avl_nodes[j]);
+        index = b * (int64_t) result->num_nodes + a;
+        CU_ASSERT_EQUAL(tsk_avl_tree_int_search(&result->pair_map, index), NULL);
+    }
+
+    ret = tsk_ibd_result_get_items(result, pairs2, lists);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    for (j = 0; j < num_pairs; j++) {
+        CU_ASSERT_EQUAL_FATAL(pairs[2 * j], pairs2[2 * j]);
+        CU_ASSERT_EQUAL_FATAL(pairs[2 * j + 1], pairs2[2 * j + 1]);
+        verify_ibd_segment_list(lists[j], result->num_nodes);
+        total_segments += lists[j]->num_segments;
+    }
+    CU_ASSERT_EQUAL_FATAL(result->total_segments, total_segments);
```

Comment:
I'm a bit unclear on what this is testing for. Is this some way of checking that there are only keys for pairs with non-empty sets of IBD segments or something?

Reply:
It's mainly check that the "keys" we return are correct. Just making sure that the two different ways we have of getting the same thing (they pairs) are the same.